examples/src/lib.rs

```diff
@@ -173,32 +180,42 @@ pub struct ExampleBase {
 
 impl ExampleBase {
     pub fn render_loop<F: Fn()>(&self, f: F) {
-        use winit::*;
-        self.events_loop.borrow_mut().run_forever(|event| {
-            f();
-            match event {
-                Event::WindowEvent { event, .. } => match event {
-                    WindowEvent::KeyboardInput { input, .. } => {
-                        if let Some(VirtualKeyCode::Escape) = input.virtual_keycode {
-                            ControlFlow::Break
-                        } else {
-                            ControlFlow::Continue
-                        }
-                    }
-                    WindowEvent::CloseRequested => winit::ControlFlow::Break,
-                    _ => ControlFlow::Continue,
-                },
-                _ => ControlFlow::Continue,
-            }
-        });
+        self.events_loop
+            .borrow_mut()
+            .run_return(|event, _, control_flow| {
+                *control_flow = ControlFlow::Wait;
+                f();
+                match event {
+                    Event::WindowEvent {
+                        event: WindowEvent::CloseRequested,
+                        window_id,
+                    } if window_id == self.window.id() => *control_flow = ControlFlow::Exit,
+
+                    Event::DeviceEvent { event, .. } => match event {
+                        DeviceEvent::Key(KeyboardInput {
+                            virtual_keycode: Some(keycode),
+                            state,
+                            ..
+                        }) => match (keycode, state) {
+                            (VirtualKeyCode::Escape, ElementState::Released) => {
+                                *control_flow = ControlFlow::Exit
+                            }
+                            _ => (),
+                        },
+                        _ => (),
+                    },
```

Comment:
We should only respond to keyboard inputs on the window, not globally, so this should be under `WindowEvent`, not `DeviceEvent`.

Reply:
You can also simplify these quite a bit by writing the expected value for `keycode` and `state` directly in the match statement. Both arms are the same, so you can combine those with `|` too. That's how I did it (without checking for `Release` state) 10 months ago:

https://github.com/Traverse-Research/ash/blob/65c6442d917cb599b18d54c5e852d5cf35aefc22/ash-window/examples/winit.rs#L37-L57

Reply:
I found this nifty I OR at https://sotrh.github.io/learn-wgpu/beginner/tutorial1-window/#the-code I hope that this will calm down clippy

Reply:
I think it's going to ask you to combine `Event::WindowEvent` and the `WindowEvent::CloseRequested | WindowEvent::KeyboardInput` though (yes, nested `|` is supported in Rust now).